Empty values and column-data-storage refactor

JorisGoosen opened the pull request:
Empty values should be stored in the data and per column.

To do this properly it required a lot of changes to the underlying code.
Ill make sure the documentation for the classes at least resembles what is going now ;)

Possible problems:
[X] importing from 0.17.3 and earlier (converting from emptyValuesMap in metadata.json)
Ignore: importing from 0.18.2 (broken emptyvalues (i might just ignore it))
[X] importing from 0.18.3 (breaking up emptyvalues into per workspace and column)
[X]] also making sure that we can still import scalar columns with a non-scalar value in them that is not convertible to a double (like in the testcsv tsv ods etc)
Should probably be a separate PR: ~~[] Add locale support (should still be done but testing will take time anyway)~~
[] importing a nominal/ordinal now as scalar means getting the ints we wrote to the _ints column. We should probably try to use the "value" shown in the label-editor instead! :o This will help with https://github.com/jasp-stats/jasp-issues/issues/2530 or https://github.com/jasp-stats/jasp-issues/issues/2325 ) (But most of that will probably be separate)

JorisGoosen commented:
<img width="1157" alt="image" src="https://github.com/jasp-stats/jasp-desktop/assets/29062713/c559ec78-9ce1-49e1-b581-c9c5eae512ac">

JorisGoosen commented:
I started a bunch of builds under `emptyValuesRefactor` in case anybode feels like testing it out.

I think this is probably ready now?

JorisGoosen commented:
Its maybe not completely done, but close enough that you guys could start reviewing it. 
And Ill also ask people to test it once the nightlies are build because im sure there are some things ive missed...

JorisGoosen commented:
[macos arm](https://static.jasp-stats.org/Nightlies/macOS-arm64/JASP-nightly-emptyValuesRefactor-420190fe0eec179bf99a022f8538946eaaf3e123.dmg)
[macos x86](https://static.jasp-stats.org/Nightlies/macOS-x86_64/JASP-nightly-emptyValuesRefactor-420190fe0eec179bf99a022f8538946eaaf3e123.dmg)

Aah, windows doesnt work it seems...

shun2wang commented:
the build test failed maybe also failing on windows.

RensDofferhoff commented:
- Press new data
- add any string in an empty column
- No Type is detected
- Use this column in descriptives and the engine crashes

JorisGoosen commented:
> * Press new data
> 
>     * add any string in an empty column
> 
>     * No Type is detected
> 
>     * Use this column in descriptives and the engine crashes

Oops! 

Edit: shouldve set columntype with the default values...
Nightlies are building with this fix for anyone curious

JorisGoosen commented:
Aah moving labels also doesnt work anymore for columns where all values are convertible to double

boutinb commented:
Some issues:

- The empty values are empty for a column, though the workspace has some empty values
- In Edit mode, if I change a value, and press the tab key, the change is lost
- With a nominal column, if I set a new value, it does not appear in the Label Editor.
- I have tried to add a computed column with just contNormal + 1: the engine crashes

Nice feature to see the labels for a scale column (if they are not the same as the value). It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.

boutinb commented on CommonData/dataset.h:
```diff
@@ -75,30 +74,27 @@ class DataSet : public DataSetBaseNode
 							*	_filtersNode			= nullptr;
 	Columns						_columns;
 	Filter					*	_filter					= nullptr;
+	EmptyValues				*	_emptyValues			= nullptr;
```
Why a pointer?
Also in Column.h: why makes EmptyValues a pointer member? Maybe because of the EmptyValues _parent initialization? By making a const pointer (it should be a const pointer), then it can be initialized.
I can add a commit for this.

JorisGoosen replied:
> Maybe because of the EmptyValues _parent initialization

Yes

> it should be a const pointer

Ok sure

boutinb commented on CommonData/emptyvalues.cpp:
```diff
@@ -1,211 +1,77 @@
+	_emptyStrings	= values;
+	_emptyDoubles	= ColumnUtils::getDoubleValues(values);
+	_hasEmptyValues	= _hasEmptyValues || _emptyStrings.size(); //Dont set it to false if it is true, and set it to true if there are empty strings set at all
 }
```
Why _hasEmptyValues should stay true if setEmptyValues is called with a an empty set of values?

JorisGoosen replied:
Because you maybe want to remove all emptyvalues for a column...

JorisGoosen commented:
> It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.

boutinb reviewed (changes requested):
I have started to review it, it will take some times.... But already here are small comments.

boutinb commented on CommonData/emptyvalues.cpp:
```diff
@@ -1,211 +1,84 @@
-		result.append(str);
-
-	return result;
+	return !_parent || _hasEmptyValues;
```
If it has a parent, it should have a look at its hasEmptyValues(), no?

JorisGoosen replied:
No, because that is always the workspace, thus always "has emptyvalues"

boutinb commented:
If you set a column to use custom empty values, delete all empty values, and save the JASP file, then when reloading the JASP file, the 'Use custom values' for this column is unset. This is because this setting is not saved in the database, and you just try to find out whether a column uses custom empty values by checking whether it has some empty values.
The settings 'Use custom values' must be saved in the database.

JorisGoosen commented:
Ooh yeah oops

boutinb commented:
If you add a new column, you don't get the Label Editor in the Variables Window.

boutinb commented:
In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.
Also if you change the label of the debinf column (in debug.csv), it crashes in column.cpp line 1234, because apparently the ColumnUtils::getDoubleValue fails.

boutinb commented on CommonData/column.cpp:
```diff
@@ -1503,372 +974,416 @@ strintmap Column::labelsSyncStrings(const stringvec &new_values, const strstrmap
+	for(double dbl : dbls)
+		if(!doubleIntIdMap.count(dbl))
+		{
+			assert(!std::isnan(dbl)); //Because why would it be?
```
How can you be sure that the label does not exist yet?

JorisGoosen replied:
We are in the function "replaceDoubleWithLabel", where we pass the dbls onwards from another function.
However, it makes no sense to replace "nan" in dbls with a label, because that is the value that would be written there if the originalValue is not convertible to double... 

So that wouldnt fit the data at all.
To enforce this and show any problems as soon as possible without confusing behaviour (given that `std::nan != std::nan` im not sure if it could ever be: `doubleIntIdMap.count(std::nan) > 0`)

JorisGoosen commented:
> In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.

Uhm, why?
That would mess up recoding quite a bit...
Suppose youve got:
```
a	| a
b	| b
c	| c
```
And you want to recode it to:
```
1	| a
2	| b
3	| c
```

Then your proposed change would make that quite annoying

shun2wang commented:
>In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.

well, I think so. because change a old value means "new value", so new value's label should be same value showing in cell. I guess in SPSS it's also same behavior, cells will showing value itself in this situation.

boutinb commented:
> > In the Label Editor, if the Label and Value have the same value, when I change the Value I expect the Label to get automatically the same value.
> 
> Uhm, why? That would mess up recoding quite a bit... Suppose youve got:
> 
> ```
> a	| a
> b	| b
> c	| c
> ```
> 
> And you want to recode it to:
> 
> ```
> 1	| a
> 2	| b
> 3	| c
> ```
> 
> Then your proposed change would make that quite annoying

I think a common case would be that if the values are 1, 2, 3..., then people wants to recode to 0, 1, 2... or 10, 20, 30....
Then by recoding them, you don't see the recoded values in the data panel, because they get their old values as labels. Quite annoying I think

JorisGoosen commented:
Maybe I could just replace the label if it is the same integer/double value and otherwise not.

boutinb commented:
Maybe one solution would be to have an empty label if the value is an integer or a double value, in this case the displayed value in the data view will be the value, and if a value is changed (recoded), then no label is created.
For string values, it makes sense to have immediately labels.

JorisGoosen commented:
Yeah but then once youve gone to a string label you cant go back anyway, so its the same thing.

boutinb commented:
When several columns are selected, it would be nice to have the background color of the headers of all selected columns to be darker. Now only one column header is darker:
![Screenshot 2024-03-14 at 16 50 39](https://github.com/jasp-stats/jasp-desktop/assets/503332/e33205fd-f023-428c-928c-7f3e90827cea)

JorisGoosen commented:
Ive fixed the value changing when the label is the same and a number.

> darker color header selected columns

Good idea

Also there are still some things wrong with the labeleditor dissappearing once there are no more labels and then not coming back.

But at least for `debInf` you now dont get a crash.
But if you put `∞` as a missing value you can make the labeleditor dissappeart from it and it wont come back :p

JorisGoosen commented:
New nightlies are building but I guess the windows one will fail again...

boutinb commented:
Also when the missing values are not custom for a column, you don't see which missing values will be used. It would be nice to see the workspace empty values in grey in place of a empty list:
![Screenshot 2024-03-14 at 16 57 02](https://github.com/jasp-stats/jasp-desktop/assets/503332/232178d2-ac08-46b5-8f20-b3d482bb219f)

boutinb commented:
> Nice feature to see the labels for a scale column (if they are not the same as the value). It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.

Don't you think that would be nice also?
In fact I think I would always set the Label as displayed value when exists, and the original value in grey (if not the same as the label)

boutinb commented:
If you open the Label Editor for a nominal column, and change one label: if you undo it, then is will undo many dummy things like "undo set name from 'contBinom' to 'contBinom'". Then it will really undo the label change. If you then redo it, it will crash (or sometimes add an extra label). This was working in. 0.18.3

shun2wang commented:
> New nightlies are building but I guess the windows one will fail again...

I can build it local on Windows but Linux still failing.

JorisGoosen commented:
> > Nice feature to see the labels for a scale column (if they are not the same as the value). It could be also nice to see the value of nominal or ordinal column, if the label and the value is not the same.
> 
> Don't you think that would be nice also? 

Maybe, it might just make the interface a bit cluttered. If you need to see the values then you should switch to scalar.

> In fact I think I would always set the Label as displayed value when exists, and the original value in grey (if not the same as the label)

I disagree, because the data you get for a scalar column is the *values* and not the *labels*, while for ordinal/nominal its the *labels*.
And this is also why I think the value shouldnt be shown for ordinal/nominal.

JorisGoosen commented:
@RensDofferhoff  @shun2wang I think linux should now be working?

shun2wang commented:
Yes, it can be build on my local now.

JorisGoosen commented:
Hmm, cut-copy-pasting with labels different from the values (see attached file, rename to `.sav`) 
[SPPS_Debug.zip](https://github.com/jasp-stats/jasp-desktop/files/14695665/SPPS_Debug.zip)

boutinb commented:
One bug: in contBinom, set custom empty values, add 0 as empty value: you see the 0's as empty values with a 0 in grey at the right hand-side. But in Edit mode, if you click on one of these empty value and leave this cell the '0' in grey is removed. If you then remove 0 as empty value, the cell stays as an empty value. Apparently just by editing the cell, it has removed the value of its empty value.

boutinb commented on Desktop/data/datasettablemodel.cpp:
```diff
@@ -25,7 +25,7 @@ DataSetTableModel::DataSetTableModel(bool showInactive)
 	
 	connect(DataSetPackage::pkg(),	&DataSetPackage::columnsFilteredCountChanged,	this, &DataSetTableModel::columnsFilteredCountChanged	);
-	connect(DataSetPackage::pkg(),	&DataSetPackage::columnDataTypeChanged,			this, [&](QString colName) { emit columnTypeChanged(colName, int(DataSetPackage::pkg()->getColumnType(colName)));	});
+	connect(DataSetPackage::pkg(),	&DataSetPackage::columnDataTypeChanged,			this, [&](QString colName) { emit columnTypeChanged(colName, int(DataSetPackage::pkg()->getColumnType(colName)));	}, Qt::QueuedConnection);
```
Quite strange to set a QueuedConnection here

boutinb reviewed (approved):
To review all the changes is quite hard...
But after some hours I think this code can be merged.

RensDofferhoff commented:
I have not found any additional bugs right now.
Might be good to merge and get the related nightlies out for wider testing if you are happy with it @JorisGoosen .

JorisGoosen commented on Desktop/data/undostack.cpp:
```diff
@@ -19,7 +19,12 @@ UndoStack::UndoStack(QObject* parent) : QUndoStack(parent)
-		push(command);
+	{
+		if (_busy)
+			Log::log() << "Try to push a new command (\"" << command->text() << "\") in the undostack while the undo or redo is not yet finished!!!" << std::endl;
```
But arent we gonna miss errors now...

Shouldnt this be an `assert`

JorisGoosen replied:
In fact, I dont like this busy thing. Why do we need this?

boutinb replied:
This is a mistake, but not a serious one: the push in the undostack is not done, but it should not have occurred: probably a command leads to a refresh that leads to trigger the same command, because no check was done whether a value was really updated (as it was the mistake this commit solves). To make JASP crashing for such a mistake is quite hard I find.

JorisGoosen replied:
that is not actually a reason, yopu are describing the cause of the crash correctly though:
- we were not checking whether useCustom was on already which added superfluous undostuff which crashes jasp
- same for emptyValues setting

I just fixed that.

After that we do not need `_busy` for anything at all.
In fact, it would just let us miss potential problems by avoiding the crash through "_busy".
And who knows what in the future might be done with this code?
Adding extra completely unecessary checks and booleans around that does not make any sense to me...

boutinb commented:
Oh loading a 0.18.3 JASP file crashes: the engine crashes when reading the emptyValues. They are apparently not upgraded in the database.

boutinb commented:
Also if I start JASP 0.16.3, add 'control' as empty value, load the debug.csv file and save it as JASP file, if I load this file in 0.19, then all the values of facExperim are 'control' and set as empty: the 'experiment' values are gone.

JorisGoosen commented:
> Oh loading a 0.18.3 JASP file crashes: the engine crashes when reading the emptyValues. They are apparently not upgraded in the database.

I tested this before and it worked. It would have been nice if you uploaded the file here...

Edit: Ah well it seems it broke again :p

JorisGoosen commented:
And its fixed!

So if @boutinb can upload the broken file from 0.16.3 I can use that (my internet is very slow and I dont really want to waste my drivespace on 0.16.3)

In the meantime im going to try to fix internal copying of labels in JASP

JorisGoosen commented:
Ok, copying labels now also works within JASP.

Ill turn on some more builds and start running some local data library tests

JorisGoosen commented:
https://static.jasp-stats.org/Nightlies/macOS-arm64/JASP-nightly-emptyValuesRefactor-da7d451794d03cb921be04f6cbc4c0725c4be3a1.dmg
https://static.jasp-stats.org/Nightlies/macOS-x86_64/JASP-nightly-emptyValuesRefactor-da7d451794d03cb921be04f6cbc4c0725c4be3a1.dmg
https://static.jasp-stats.org/Nightlies/WindowsZip/JASP-nightly-emptyValuesRefactor-da7d451794d03cb921be04f6cbc4c0725c4be3a1.zip

shun2wang commented:
Test feedback( on windows 11 with portable .zip):
1. load new .sav dataset will open the filter window by default?
2.  Long press<kbd>Shift</kbd>+cursor for multiple selection :selection range behavior inconsistent when selecting forward /up and backward/down. the selected range should alway 'append' but not sliding.
3. Compute new variable crash engine:
```
00:08:32.419: EngineRepresentation::handleEngineCrash():
**Engine # 0** process *crashed* and it's state is *computeColumn*.
00:08:32.423: R_HOME set to D:/Program/JASP/JASP-N~4/R
00:08:32.423: Windows PATH was changed to: 'D:/Program/JASP/JASP-nightly-emptyValuesRefactor;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin/x64'
00:08:32.508: informing engine #0 that it ought to restart
00:08:32.508: Setting new engine process to engineRepresentation Engine #0
00:08:32.546: informing engine #0 that it may resume.
00:08:32.546: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.550: Received QML warnings:
	qrc:/components/JASP/Controls/ComboBox.qml:129: TypeError: Value is undefined and could not be converted to an object

00:08:32.589: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.622: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.683: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.757: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.818: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.882: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.939: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:32.998: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.059: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.135: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.192: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.255: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.325: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.374: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.431: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.510: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.573: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.628: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.689: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.744: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.808: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.883: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.937: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:33.999: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.060: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.125: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.185: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.258: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.309: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.372: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.439: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.522: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.562: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.633: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.693: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.743: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.813: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.874: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.934: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:34.996: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.070: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.131: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.190: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.249: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.310: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.373: EngineRepresentation::processEngineResumedReply() for engine #0
00:08:35.374: EngineRepresentation::sendSettings()
00:08:35.374: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:35.435: EngineRepresentation::sendLogCfg()
00:08:35.436: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:39.647: Engine # 0 finished while in state 'computeColumn' crashing and with exitCode -1073740791!
00:08:39.647: EngineRepresentation::handleEngineCrash():
**Engine # 0** process *crashed* and it's state is *computeColumn*.
00:08:39.650: R_HOME set to D:/Program/JASP/JASP-N~4/R
00:08:39.650: Windows PATH was changed to: 'D:/Program/JASP/JASP-nightly-emptyValuesRefactor;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin/x64'
00:08:39.716: informing engine #0 that it ought to restart
00:08:39.716: Setting new engine process to engineRepresentation Engine #0
00:08:39.727: informing engine #0 that it may resume.
00:08:42.564: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:42.632: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:42.693: EngineRepresentation::processEngineResumedReply() for engine #0
00:08:42.693: EngineRepresentation::sendSettings()
00:08:42.694: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:42.754: EngineRepresentation::sendLogCfg()
00:08:42.755: Not enough idle engines! Need  one for compcol, one will be idle soon...
00:08:46.194: Engine # 0 finished while in state 'computeColumn' crashing and with exitCode -1073740791!
00:08:46.194: EngineRepresentation::handleEngineCrash():
**Engine # 0** process *crashed* and it's state is *computeColumn*.
00:08:46.198: R_HOME set to D:/Program/JASP/JASP-N~4/R
00:08:46.198: Windows PATH was changed to: 'D:/Program/JASP/JASP-nightly-emptyValuesRefactor;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin;D:/Program/JASP/JASP-nightly-emptyValuesRefactor/R/bin/x64'
00:08:46.262: informing engine #0 that it ought to restart
00:08:46.262: Setting new engine process to engineRepresentation Engine #0
00:08:46.283: informing engine #0 that it may resume.
00:08:49.441: EngineRepresentation::processEngineResumedReply() for engine #0
00:08:49.441: EngineRepresentation::sendSettings()
00:08:49.494: EngineRepresentation::sendLogCfg()
```

shun2wang commented on Desktop/components/JASP/Widgets/LabelEditorWindow.qml:
```diff
@@ -135,7 +135,7 @@ FocusScope
 							anchors.fill:		selectionRectangle
 							acceptedButtons:	Qt.LeftButton
-							cursorShape:		Qt.PointingHandCursor
+							cursorShape:		Qt.CrossCursor	
```
I've always wanted to have a consistent mouse shape on the cell pointer in all table views, but that's not important:-)

JorisGoosen replied:
Well, this one should be only for "selecting cells" and to distinguish from the textfield/filter-checkbox.
Id prefer a better cursor than this cross but qt doesnt offer it...

JorisGoosen committed:
ok lets clean the emptyvalues mess up

why do we need all that missingdata ?

ooh my god so much shit

postmerge fix en more stuff removed

continue with setToNominalText etc

it compiles and crashes...

ok it compiles and it runs

but scalar are not visible (all missing) and emptyvalues are ignored

the definition of the database is now in its own file though
so we can just look at pure SQL that will get included without needing Qts resourcesystem

this is more like it

aaah copy constructors..

ordinal missingvalues show up again in dataviewer

ik wil emptyvalues ook aan de engine kant maar dit sloopt alles

maybe less silly declaration of int missing value

column empty values are now used by R but not yet workspace

apply changes also in R

better name

:'(

give back original value as double instead of entry in _ints

should help out with recoding variables

intermediate commit

remove more!

more refactoring

more changes

simplify data reading in/to R

less crashy

pannekoeken

data now actually arrives in R

this works but empty values on a column is broken

for instance, set empty value "1" or something on contBinom and then see broken factor

empty values are now passed to R correctly

why use the length if you could use the variant directly?

Column now also shows double values etc as labels+value

not editable yet though

make sure a label is created when someone tries to change something for a double label in the variableswindow

filterAllows is broken, just like setting value and perhaps even description?

hmm

allow editing both label and original value

this should make the value updsate with originalValue changes

recoding values now works

Try to make the dataviewer update after type change because certain values can be different then

cosmetics

remove new yet unnecessary code

replace all labels upto edited one

the row for the label should not be looked up in _dbls

labels moven en reversen also works now

Please give a compile error if I forget a return

MASSIVE speed up

Tested with CSVs of 200k and 1 million rows ^^

cleaning up some more

fix datetimeprintfunction and silence 2 warnings

more code removed and loading from sav file now seems flawless

remove some more (probably) uneccessary stuff

Not supersure about the old import still working, but ill test that later

handle edits better

some colortweak and rename computedcolumnsmodel to computedcolumn

the actual changes for columnmodel and attempts to make the width corectly for the data columns

better handling of value==""

properly determine unicode column width

make sure default description() in importdataset doesnt mess things up

show labels for scale column

change multiple columntype at once

Implements https://github.com/jasp-stats/jasp-issues/issues/2065

make sure columnheaders update and keep updating

Also rewrote the reseting of columntypes to use new stuff

store result of qprocess debug info in static to make sure we never ever run it more than once

add support for loading older jasp files (seems to work with 0.14 and up at least)

also loads missing data properly now

make sure aboutmodel only looks debug stuff up when necessary!

support reading from 0.18.3 files

filterallow should work on doubles

make sure the column type is changed of the correct column (at least the one where you click the menu!)

at least msvc always gives an error for non-void returning nothing

Write doubles on change as doubles, not ints..

set columntype when setting default values on them

fix moving labels

remove superfluous row* functions from doc on rowfuncs

ignore decimal stuff in doubles completely
comma dot all fine

add some documentation to Column that reflects the current state

add some documentation to Column

remove some stuff

make sure that emptyvalues of column get initialised properly when selecting "use custom empty values"

const pointers for bruno

slightly better handling ovrf computed columns

makes interacting with labeleditor smoother by only doing all the work for actual changes...

fix "use custom empty values" not getting saved when there are 0 missing values

make labeleditor bit better behaved

filterallows works better now,
things get selected when edited

double conversion should be consistent

try to keep int and double value+label synchronized

make sure description in workspace settings wraps

remove useless vars

make changing values smooth and slippery

show workspace emptyValues when no custom column empty values

fix column-undos

show all selected row and columnheaders highlighted

show label for emptyvvalues as mentioned at https://github.com/jasp-stats/jasp-issues/issues/2638#issuecomment-2003683879

Also adds a better Ctrl/Cmd+A handling in dataview:
- first selects cell
- second time selects all cells

use jaspBase without test error and fix non-void return missed because gcc is a bit slow

clean up a bit in ColumnModel::computedTypeValues

better behaved selection and less logging

DataTableView/DataSetView:
make sure that when you open the rightclickmenu in a selection (cut for instance) it uses the topleft corner of the selection.
dont reset the selection when opening a rightclickmenu in a selection
selections now persist beyond datasetpackage reset
copy/cut/paste added as keyhandler on DataTableView
disabled caching of items by default, still turned on for normal dataviewing though

small fix to search in results, it should only try to use resultsView if it actually exists

cutting stuff shouldnt drop all labels

return something from the procesinfo if no parent

add  all sql as normal files to librry

make sure to check whether useEmptyValues is on or emptyValues are the same to avoid undo crash

also make sure that columntypes, computedcolumntypes and tabs in the variableswindow are immediately translated instead of only after reboot

support reading from 0.18.3 again

make sure <0.18 files still work

(avoid the 1-based nonsense in intsId)

support copying labels internally in jasp

make sure nonvoid returns something

fix broken Leukemia file in survival data library

JorisGoosen committed:
manually added columns should open variableswindow for convenience

JorisGoosen commented:
> Test feedback( on windows 11 with portable .zip):

Thanks ^^

>     1. load new .sav dataset will open the filter window by default?

I havent seen this with my test case but maybe it does for yours?
Can you upload it?

>     2. Long pressShift+cursor for multiple selection :selection range behavior inconsistent when selecting forward /up and backward/down. the selected range should alway 'append' but not sliding.

The only inconsistency I see when using shift+cursor is that going left you first have to go through the text in the edititem.

>     3. Compute new variable crash engine:
What exactly did you do here?

boutinb committed:
Do not remove custom empty value when editing cell

If a cell has a custom empty value, in a Edit mode when clicking on this cell, this removes the custom empty value.
For example, add 1 as empty value for the facFive column. Click on a cell that has 1 as value (it is displayed in grey), and go to another cell: the value 1 is removed.

boutinb commented:
If a cell has a custom empty value, in a Edit mode when clicking on this cell, this removes the custom empty value.
For example, add 1 as empty value for the facFive column. Click on a cell that has 1 as value (it is displayed in grey), and go to another cell: the value 1 is removed.
I have just added a commit for this.

shun2wang commented:
for 1 and 3,Sorry, in fact it can only be reproduced when I run it for the first time using .zip. After that I restart JASP and it cannot be reproduced again. I will report it separately once I find a reproducible path.

for 2, I mean row<up/down> col <left/right> will produce selection sliding behavior,see:
[screenshot_20240327.webm](https://github.com/jasp-stats/jasp-desktop/assets/10348402/e335f75c-d3cc-4742-823b-e45bde208402) press the Shift and click left/up direction will get sliding selection.

JorisGoosen committed:
add computed column insertion back into menus

also makes sure to show the variableswindow and thus code-editor on insertion only for computed columns.
This to allow for rapid addition of computed columns

Also refactors a bit of the menu code in DataTableView

JorisGoosen commented:
Ah, yeah now I see, that is definitely weird behaviour :p

JorisGoosen committed:
improve selection behaviour

JorisGoosen commented:
Ok, ive fixed the selection behaviour (and improved it ^^)

JorisGoosen committed:
fix a few problems and remove some stuff

draghand cursor for cells ;)
use same double formatting for label
undo for label originalvalue should remember label and set it back

JorisGoosen committed:
Dont try to write to isComputed in Columns

JorisGoosen committed:
dont just break on asking computed columns without data

fix Process data library file so it will actually load

JorisGoosen committed:
add a testfile, generated in 0.18.3 with Descriptives run on debug.csv columns

JorisGoosen committed:
dont crash on actual missing value value

JorisGoosen commented:
Testing the data library revealed some bugs, that I then proceeded to fix, like:
- opening a jasp file, closing it and opening another one caused the load to fail...
- an actual missing value in the data would crash the engine on loading that column as ordinal/nominal
- a process data library file was not saved correctly. I overwrote it with the one from 0.18.3
- added a testfile with descriptives run on all columns in 0.18.3 debug.csv. Maximum statistics on in all tables (but not everything cause descriptives could not in fact handle that...)

Reloading the testfile with the emptyValuesRefactor build showed that the behaviour in Descriptives of what used to be Nominal text is hampered a bit. But im sure that requires only a light touch on the descriptives R code side

JorisGoosen committed:
show value for ordinal/nominal as "shadow display"

As requested by Bruno in the PR
And by Thomas in https://github.com/jasp-stats/jasp-issues/issues/1786

JorisGoosen committed:
small tweaks

set some code font in a few places
dont set columntype to unknown when running setDefaultValues
add a codepath for forcing all columns to the same columntype for computed columns

JorisGoosen committed:
tweak darktheme colors a bit and apply jasp UI font everywhere...

JorisGoosen committed:
add support for forcing input types of computed columns

JorisGoosen committed:
make sure all dropdowns in prefsUI have the same width...

shun2wang commented:
FileOperation menu has some overlap with the `▶`  , especially for language-specific strings lengths.

![image](https://github.com/jasp-stats/jasp-desktop/assets/10348402/47c14837-a8aa-47e2-98c8-18b8e230fa3e)

JorisGoosen committed:
Make sure menu arrow is on the right entirely

JorisGoosen commented:
Ive also added a new feature to computed columns, see: https://github.com/jasp-stats/INTERNAL-jasp/issues/2487

JorisGoosen committed:
make sure missing value shadowtext is also shown for scalar

JorisGoosen committed:
make sure NA doesnt get written to nominal/ordinal columns for completely empty scalar cells

also malkes sure the keep and convert input columntypes button always actually shows the correct value

boutinb commented:
When reading a JASP file from 0.18.0 where for example 0 was set as an empty value, JASP displays the values that had 0 as empty values in the data view, but does not display the original value (0) in grey, and it does not set in the workspace that 0 is an empty value.

Sorry bad manipulation. It works: the 0 is set as empty value in the workspace. Only minor thing is that the if 0 is unset, the cells are not set back to 0: the original value (0) is lost.

boutinb commented:
> Sorry bad manipulation. It works: the 0 is set as empty value in the workspace. Only minor thing is that the if 0 is unset, the cells are not set back to 0: the original value (0) is lost.

I have built it with the last commits, and now if I load a JASP file from 0.18.0, the workspace empty values are empty!

JorisGoosen commented:
I only made an effort to support pre-0.18.0 versions of JASP and specifically 0.18.3.
This because there were some quite terrible bugs in those earlier 0.18.* versions and because we released them in such short order I think we can safely ignore them.
